File: backend/app/storage/local.py
"""Local-filesystem object store + HMAC-signed URL helpers.

Replaces the three former Supabase Storage buckets
(``meeting-recordings``, ``deal-documents``, ``deliverables``) with files on
the worker's local disk under ``settings.storage_root/{bucket}/{key}``.

Access is gated by short-lived HMAC-SHA256 signatures issued by the worker:
the signature over ``"{bucket}:{key}:{expires_at}"`` IS the capability to
read or write that object until it expires.
"""

from __future__ import annotations

import hashlib
import hmac
import time
from pathlib import Path

from app.core.config import settings

BUCKETS = {"meeting-recordings", "deal-documents", "deliverables"}


def _root() -> Path:
    return Path(settings.storage_root)


def _safe_path(bucket: str, key: str) -> Path:
    """Resolve ``{root}/{bucket}/{key}``, guaranteeing the result stays inside
    the bucket directory (defends against path traversal via the key)."""
    if bucket not in BUCKETS:
        raise ValueError(f"unknown bucket: {bucket!r}")
    if key.startswith("/"):
        raise ValueError(f"unsafe key: {key!r}")
    bucket_root = (_root() / bucket).resolve()
    candidate = (bucket_root / key).resolve()
    if not candidate.is_relative_to(bucket_root):
        raise ValueError(f"unsafe key: {key!r}")
    return candidate


# ---- object operations ----------------------------------------------------
def save_bytes(bucket: str, key: str, data: bytes) -> None:
    path = _safe_path(bucket, key)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(data)


def read_bytes(bucket: str, key: str) -> bytes:
    return _safe_path(bucket, key).read_bytes()


def object_path(bucket: str, key: str) -> Path:
    """Validated (traversal-safe) filesystem path for an object.

    Lets callers stream a file straight off disk (e.g. ``FileResponse``)
    instead of buffering the whole thing into memory via :func:`read_bytes`.
    """
    return _safe_path(bucket, key)


def delete(bucket: str, key: str) -> None:
    path = _safe_path(bucket, key)
    if path.exists():
        path.unlink()


def exists(bucket: str, key: str) -> bool:
    return _safe_path(bucket, key).is_file()


# ---- URL signing ----------------------------------------------------------
def _signing_key() -> str:
    if settings.storage_signing_key:
        return settings.storage_signing_key
    # Dev/test convenience only — production boot requires STORAGE_SIGNING_KEY
    # (config._require_prod_secrets), so storage signing never silently reuses
    # the shared internal token in prod.
    if not settings.is_production:
        return settings.worker_internal_token
    raise RuntimeError("STORAGE_SIGNING_KEY is not configured")


def sign(method: str, bucket: str, key: str, expires_at: int) -> str:
    # Bind the HTTP method into the signature so a GET (download) URL can't be
    # replayed as a PUT (overwrite) and vice-versa.
    msg = f"{method.upper()}:{bucket}:{key}:{expires_at}".encode()
    return hmac.new(_signing_key().encode(), msg, hashlib.sha256).hexdigest()


def make_signed_url(
    bucket: str, key: str, *, method: str = "GET", ttl_seconds: int = 3600
) -> str:
    exp = int(time.time()) + ttl_seconds
    sig = sign(method, bucket, key, exp)
    return f"/api/v1/storage/{bucket}/{key}?expires={exp}&sig={sig}"


def verify(method: str, bucket: str, key: str, expires_at: int, sig: str) -> bool:
    if int(time.time()) > expires_at:
        return False
    expected = sign(method, bucket, key, expires_at)
    return hmac.compare_digest(expected, sig)

File: worker/internal/storage/local.go
// Package storage is the local-filesystem object store + HMAC-signed URL helpers
// (ports app/storage/local.py). Files live under {root}/{bucket}/{key}; a
// short-lived HMAC-SHA256 signature over "METHOD:bucket:key:expires" IS the
// capability to read/write that object.
package storage

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Buckets that exist (deliverables are produced server-side).
var Buckets = map[string]bool{
	"meeting-recordings": true,
	"deal-documents":     true,
	"deliverables":       true,
}

var (
	ErrUnknownBucket = errors.New("unknown bucket")
	ErrUnsafeKey     = errors.New("unsafe key")
)

// safePath resolves {root}/{bucket}/{key}, guaranteeing the result stays inside
// the bucket dir (defends against path traversal via the key).
func safePath(root, bucket, key string) (string, error) {
	if !Buckets[bucket] {
		return "", ErrUnknownBucket
	}
	if strings.HasPrefix(key, "/") {
		return "", ErrUnsafeKey
	}
	bucketRoot, err := filepath.Abs(filepath.Join(root, bucket))
	if err != nil {
		return "", err
	}
	candidate, err := filepath.Abs(filepath.Join(bucketRoot, filepath.FromSlash(key)))
	if err != nil {
		return "", err
	}
	if candidate != bucketRoot && !strings.HasPrefix(candidate, bucketRoot+string(os.PathSeparator)) {
		return "", ErrUnsafeKey
	}
	return candidate, nil
}

// SaveBytes writes data to {root}/{bucket}/{key}, creating parent dirs.
func SaveBytes(root, bucket, key string, data []byte) error {
	p, err := safePath(root, bucket, key)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return err
	}
	return os.WriteFile(p, data, 0o644)
}

// ObjectPath returns the traversal-safe on-disk path (for streaming downloads).
func ObjectPath(root, bucket, key string) (string, error) {
	return safePath(root, bucket, key)
}

// Exists reports whether the object is a regular file.
func Exists(root, bucket, key string) bool {
	p, err := safePath(root, bucket, key)
	if err != nil {
		return false
	}
	fi, err := os.Stat(p)
	return err == nil && fi.Mode().IsRegular()
}

// sign computes the method-bound HMAC-SHA256 hex signature.
func sign(signingKey, method, bucket, key string, expiresAt int64) string {
	msg := fmt.Sprintf("%s:%s:%s:%d", strings.ToUpper(method), bucket, key, expiresAt)
	mac := hmac.New(sha256.New, []byte(signingKey))
	mac.Write([]byte(msg))
	return hex.EncodeToString(mac.Sum(nil))
}

// MakeSignedURL returns a relative capability URL valid for ttl seconds.
func MakeSignedURL(signingKey, bucket, key, method string, ttl time.Duration) string {
	exp := time.Now().Unix() + int64(ttl.Seconds())
	sig := sign(signingKey, method, bucket, key, exp)
	return fmt.Sprintf("/api/v1/storage/%s/%s?expires=%d&sig=%s", bucket, key, exp, sig)
}

// Verify checks expiry + the method-bound signature (constant-time).
func Verify(signingKey, method, bucket, key string, expiresAt int64, sig string) bool {
	if time.Now().Unix() > expiresAt {
		return false
	}
	expected := sign(signingKey, method, bucket, key, expiresAt)
	return subtle.ConstantTimeCompare([]byte(expected), []byte(sig)) == 1
}
